### ISSUE TITLE
Pytorch reimplemention has something wrong!

### ISSUE THREAD

**Issue by goodboyyes2009** [jaymody/picoGPT#22]
Referring to your code, I [reimplemented it using torch](https://github.com/goodboyyes2009/picoGPT). but there are some problems. Can you help me take a look?

**Comment by jaymody** [comment]
The problem is [in the `forward` function of `TransformerBlock`](https://github.com/goodboyyes2009/picoGPT/blob/d21a4c2d3b00c167598f61250e8518e7335e4617/gpt2_torch.py#L41-L47).

```python
    def forward(self, x):
        x = self.norm1(x)
        x = x + self.mulit_head_attention(x)
        x = self.norm2(x)
        # dump("attn_x_torch.pkl", self.ffn(x).detach().cpu().numpy())
        x = x + self.ffn(x)
        return x
```

The issue here is that you apply normalization to `x` and reassign that to `x` itself, meaning when you do `x + mha(x)` and `x + ffn(x)` both the `x`'s have normalization applied to them when ONLY the one being fed to `mha`/`ffn` should be normalized. See:

```python
def transformer_block(x, mlp, attn, ln_1, ln_2, n_head):
    x = x + mha(layer_norm(x, **ln_1), **attn, n_head=n_head)
    x = x + ffn(layer_norm(x, **ln_2), **mlp)
    return x
```

Notice how layer norm is only applied to the input of `mha`/`ffn` and not the residual connection.

So, the fix is simple:

```python
    def forward(self, x):
        x_norm = self.norm1(x)
        x = x + self.mulit_head_attention(x_norm)
        x_norm = self.norm2(x)
        x = x + self.ffn(x_norm)
        return x
```

After this change, I obtained the correct output:

```
$ python gpt2_torch.py "Alan Turing theorized that computers would one day become" 40 "124M" "models"
generating: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 40/40 [00:01<00:00, 22.95it/s]
 the most powerful machines on the planet.

The computer is a machine that can perform complex calculations, and it can perform these calculations in a way that is very similar to the human brain.
```

I have a PyTorch implementation (including training code) [here](https://github.com/jaymody/simpleGPT/blob/main/src/simplegpt/model.py) that I think would be a good reference.

**Comment by goodboyyes2009** [comment]
Thank you very much for your guidance. @jaymody
